### DISCUSSION TITLE
find hyperparameters using optuna

### DISCUSSION THREAD

**Discussion by andysingal** [optuna/optuna#4908] [Q&A]
### Expected behavior

run the code to train and validate the model 

### Environment

- Optuna version:
- Python version:
- OS:
- (Optional) Other libraries and their versions:

### Error messages, stack traces, or logs

```shell
### 🐛 Describe the bug
for code, please check hyperparameter section of https://www.kaggle.com/code/artemmalygin/pytorch-animals-10#Find-best-hyperparameters-with-optuna

def objective(trial):
    N_TRAIN_EXAMPLES = CONFIG['train_batch_size'] * 30
    N_VALID_EXAMPLES = CONFIG['valid_batch_size'] * 10
    device = CONFIG['device']

    # Generate the model.
    model = AnimalsModel(CONFIG['model_name'], labels)

    # Generate the optimizers.
    optimizer_number = trial.suggest_categorical('optimizer number (0: Adam; 1: SGD)', [0, 1])
    lr = trial.suggest_float('learning rate', 1e-2, 1e-1, log=True)
    
    optimizer_map = {
        0: 'Adam',
        1: 'SGD',
    }
    
    optimizer = getattr(optim, optimizer_map[optimizer_number])(model.parameters(), lr=lr)
    
    config = {
        'optimizer (0: Adam; 1: SGD)': optimizer_number,
        'learning rate': lr,
    }
    
    run = wandb.init(project='Animals10',
                     name=f'trial_{trial.number + 1}',
                     group='optuna research',
                     config=config,
                     anonymous='must')

    # Training of the model.
    for epoch in range(CONFIG['epochs']):
        model.train()
    
        dataset_size = 0
        running_loss = 0.0
        
        for batch_idx, data in enumerate(train_loader):
            # Limiting training data for faster epochs.
            if batch_idx * CONFIG['train_batch_size'] > N_TRAIN_EXAMPLES:
                break

            images = data['image'].to(device, dtype=torch.float)
            targets = data['target'].to(device, dtype=torch.long)
            targets = torch.reshape(targets, (-1,))

            batch_size = images.size(0)

            outputs = model(images)
            loss = criterion(outputs, targets)
            loss.backward()
            optimizer.step()
            
            # zero the parameter gradients
            optimizer.zero_grad()

        # Validation of the model.
        model.eval()
        correct = 0
        with torch.no_grad():
            for batch_idx, data in enumerate(valid_loader):
                # Limiting validation data.
                if batch_idx * CONFIG['valid_batch_size'] > N_VALID_EXAMPLES:
                    break
                
                images = data['image'].to(device, dtype=torch.float)
                targets = data['target'].to(device, dtype=torch.long)
                targets = torch.reshape(targets, (-1,))

                batch_size = images.size(0)

                outputs = model(images)
                loss = criterion(outputs, targets)

                running_loss += (loss.item() * batch_size)
                dataset_size += batch_size

        epoch_loss = running_loss / dataset_size
        run.log({'Cross Entropy Loss': epoch_loss})

        trial.report(epoch_loss, epoch)

        # handle pruning based on the intermediate value.
        if trial.should_prune():
            raise optuna.exceptions.TrialPruned()

    run.finish()
    
    return epoch_loss
```
```
study = optuna.create_study()
study.optimize(objective, n_trials=20, timeout=1800, show_progress_bar=True)

print('Number of finished trials: {}'.format(len(study.trials)))

print('Best trial:')
trial = study.best_trial

print('  Value: {}'.format(trial.value))

print('  Params: ')
for key, value in trial.params.items():
    print('    {}: {}'.format(key, value))
          
# Create the summary run.
summary = wandb.init(project='Animals10',
                     group='optuna summary',
                     name='summary')

# Getting the study trials.
trials = study.trials

# WandB summary.
for step, trial in enumerate(trials):
    # Logging the loss.
    summary.log({'Cross Entropy Loss': trial.value}, step=step, commit=False)

    # Logging the parameters.        
    summary.log(trial.params, commit=True)
    
summary.finish()
```
error:
```
View run at https://wandb.ai/andysingal/Animals10/runs/s26ixk76
[W 2023-09-09 14:57:30,926] Trial 0 failed with parameters: {'optimizer number (0: Adam; 1: SGD)': 0, 'learning rate': 0.0646134223274061} because of the following error: TypeError('Caught TypeError in DataLoader worker process 0.\nOriginal Traceback (most recent call last):\n  File "/opt/conda/lib/python3.10/site-packages/torch/utils/data/_utils/worker.py", line 308, in _worker_loop\n    data = fetcher.fetch(index)\n  File "/opt/conda/lib/python3.10/site-packages/torch/utils/data/_utils/fetch.py", line 51, in fetch\n    data = [self.dataset[idx] for idx in possibly_batched_index]\n  File "/opt/conda/lib/python3.10/site-packages/torch/utils/data/_utils/fetch.py", line 51, in <listcomp>\n    data = [self.dataset[idx] for idx in possibly_batched_index]\n  File "/tmp/ipykernel_29/995576240.py", line 18, in __getitem__\n    image = self.transforms(image=image)[\'image\']\n  File "/opt/conda/lib/python3.10/site-packages/albumentations/core/composition.py", line 210, in __call__\n    data = t(**data)\n  File "/opt/conda/lib/python3.10/site-packages/albumentations/core/transforms_interface.py", line 118, in __call__\n    return self.apply_with_params(params, **kwargs)\n  File "/opt/conda/lib/python3.10/site-packages/albumentations/core/transforms_interface.py", line 131, in apply_with_params\n    res[key] = target_function(arg, **dict(params, **target_dependencies))\n  File "/opt/conda/lib/python3.10/site-packages/albumentations/augmentations/transforms.py", line 213, in apply\n    return F.normalize(image, self.mean, self.std, self.max_pixel_value)\n  File "/opt/conda/lib/python3.10/site-packages/albumentations/augmentations/functional.py", line 89, in normalize\n    mean = np.array(mean, dtype=np.float32)\n  File "/opt/conda/lib/python3.10/site-packages/torch/_tensor.py", line 972, in __array__\n    return self.numpy().astype(dtype, copy=False)\nTypeError: can\'t convert cuda:0 device type tensor to numpy. Use Tensor.cpu() to copy the tensor to host memory first.\n').
Traceback (most recent call last):
  File "/opt/conda/lib/python3.10/site-packages/optuna/study/_optimize.py", line 200, in _run_trial
    value_or_values = func(trial)
  File "/tmp/ipykernel_29/2342615095.py", line 38, in objective
    for batch_idx, data in enumerate(train_loader):
  File "/opt/conda/lib/python3.10/site-packages/torch/utils/data/dataloader.py", line 634, in __next__
    data = self._next_data()
  File "/opt/conda/lib/python3.10/site-packages/torch/utils/data/dataloader.py", line 1346, in _next_data
    return self._process_data(data)
  File "/opt/conda/lib/python3.10/site-packages/torch/utils/data/dataloader.py", line 1372, in _process_data
    data.reraise()
  File "/opt/conda/lib/python3.10/site-packages/torch/_utils.py", line 644, in reraise
    raise exception
TypeError: Caught TypeError in DataLoader worker process 0.
Original Traceback (most recent call last):
  File "/opt/conda/lib/python3.10/site-packages/torch/utils/data/_utils/worker.py", line 308, in _worker_loop
    data = fetcher.fetch(index)
  File "/opt/conda/lib/python3.10/site-packages/torch/utils/data/_utils/fetch.py", line 51, in fetch
    data = [self.dataset[idx] for idx in possibly_batched_index]
  File "/opt/conda/lib/python3.10/site-packages/torch/utils/data/_utils/fetch.py", line 51, in <listcomp>
    data = [self.dataset[idx] for idx in possibly_batched_index]
  File "/tmp/ipykernel_29/995576240.py", line 18, in __getitem__
    image = self.transforms(image=image)['image']
  File "/opt/conda/lib/python3.10/site-packages/albumentations/core/composition.py", line 210, in __call__
    data = t(**data)
  File "/opt/conda/lib/python3.10/site-packages/albumentations/core/transforms_interface.py", line 118, in __call__
    return self.apply_with_params(params, **kwargs)
  File "/opt/conda/lib/python3.10/site-packages/albumentations/core/transforms_interface.py", line 131, in apply_with_params
    res[key] = target_function(arg, **dict(params, **target_dependencies))
  File "/opt/conda/lib/python3.10/site-packages/albumentations/augmentations/transforms.py", line 213, in apply
    return F.normalize(image, self.mean, self.std, self.max_pixel_value)
  File "/opt/conda/lib/python3.10/site-packages/albumentations/augmentations/functional.py", line 89, in normalize
    mean = np.array(mean, dtype=np.float32)
  File "/opt/conda/lib/python3.10/site-packages/torch/_tensor.py", line 972, in __array__
    return self.numpy().astype(dtype, copy=False)
TypeError: can't convert cuda:0 device type tensor to numpy. Use Tensor.cpu() to copy the tensor to host memory first.

[W 2023-09-09 14:57:30,942] Trial 0 failed with value None.
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
Cell In[49], line 2
      1 study = optuna.create_study()
----> 2 study.optimize(objective, n_trials=20, timeout=1800, show_progress_bar=True)
      4 print('Number of finished trials: {}'.format(len(study.trials)))
      6 print('Best trial:')

File /opt/conda/lib/python3.10/site-packages/optuna/study/study.py:442, in Study.optimize(self, func, n_trials, timeout, n_jobs, catch, callbacks, gc_after_trial, show_progress_bar)
    339 def optimize(
    340     self,
    341     func: ObjectiveFuncType,
   (...)
    348     show_progress_bar: bool = False,
    349 ) -> None:
    350     """Optimize an objective function.
    351 
    352     Optimization is done by choosing a suitable set of hyperparameter values from a given
   (...)
    440             If nested invocation of this method occurs.
    441     """
--> 442     _optimize(
    443         study=self,
    444         func=func,
    445         n_trials=n_trials,
    446         timeout=timeout,
    447         n_jobs=n_jobs,
    448         catch=tuple(catch) if isinstance(catch, Iterable) else (catch,),
    449         callbacks=callbacks,
    450         gc_after_trial=gc_after_trial,
    451         show_progress_bar=show_progress_bar,
    452     )

File /opt/conda/lib/python3.10/site-packages/optuna/study/_optimize.py:66, in _optimize(study, func, n_trials, timeout, n_jobs, catch, callbacks, gc_after_trial, show_progress_bar)
     64 try:
     65     if n_jobs == 1:
---> 66         _optimize_sequential(
     67             study,
     68             func,
     69             n_trials,
     70             timeout,
     71             catch,
     72             callbacks,
     73             gc_after_trial,
     74             reseed_sampler_rng=False,
     75             time_start=None,
     76             progress_bar=progress_bar,
     77         )
     78     else:
     79         if n_jobs == -1:

File /opt/conda/lib/python3.10/site-packages/optuna/study/_optimize.py:163, in _optimize_sequential(study, func, n_trials, timeout, catch, callbacks, gc_after_trial, reseed_sampler_rng, time_start, progress_bar)
    160         break
    162 try:
--> 163     frozen_trial = _run_trial(study, func, catch)
    164 finally:
    165     # The following line mitigates memory problems that can be occurred in some
    166     # environments (e.g., services that use computing containers such as GitHub Actions).
    167     # Please refer to the following PR for further details:
    168     # [https://github.com/optuna/optuna/pull/325.](https://github.com/optuna/optuna/pull/325.%3C/span%3E)
    169     if gc_after_trial:

File /opt/conda/lib/python3.10/site-packages/optuna/study/_optimize.py:251, in _run_trial(study, func, catch)
    244         assert False, "Should not reach."
    246 if (
    247     frozen_trial.state == TrialState.FAIL
    248     and func_err is not None
    249     and not isinstance(func_err, catch)
    250 ):
--> 251     raise func_err
    252 return frozen_trial

File /opt/conda/lib/python3.10/site-packages/optuna/study/_optimize.py:200, in _run_trial(study, func, catch)
    198 with get_heartbeat_thread(trial._trial_id, study._storage):
    199     try:
--> 200         value_or_values = func(trial)
    201     except exceptions.TrialPruned as e:
    202         # TODO(mamu): Handle multi-objective cases.
    203         state = TrialState.PRUNED

Cell In[48], line 38, in objective(trial)
     35 dataset_size = 0
     36 running_loss = 0.0
---> 38 for batch_idx, data in enumerate(train_loader):
     39     # Limiting training data for faster epochs.
     40     if batch_idx * CONFIG['train_batch_size'] > N_TRAIN_EXAMPLES:
     41         break

File /opt/conda/lib/python3.10/site-packages/torch/utils/data/dataloader.py:634, in _BaseDataLoaderIter.__next__(self)
    631 if self._sampler_iter is None:
    632     # TODO([https://github.com/pytorch/pytorch/issues/76750)](https://github.com/pytorch/pytorch/issues/76750)%3C/span%3E)
    633     self._reset()  # type: ignore[call-arg]
--> 634 data = self._next_data()
    635 self._num_yielded += 1
    636 if self._dataset_kind == _DatasetKind.Iterable and \
    637         self._IterableDataset_len_called is not None and \
    638         self._num_yielded > self._IterableDataset_len_called:

File /opt/conda/lib/python3.10/site-packages/torch/utils/data/dataloader.py:1346, in _MultiProcessingDataLoaderIter._next_data(self)
   1344 else:
   1345     del self._task_info[idx]
-> 1346     return self._process_data(data)

File /opt/conda/lib/python3.10/site-packages/torch/utils/data/dataloader.py:1372, in _MultiProcessingDataLoaderIter._process_data(self, data)
   1370 self._try_put_index()
   1371 if isinstance(data, ExceptionWrapper):
-> 1372     data.reraise()
   1373 return data

File /opt/conda/lib/python3.10/site-packages/torch/_utils.py:644, in ExceptionWrapper.reraise(self)
    640 except TypeError:
    641     # If the exception takes multiple arguments, don't try to
    642     # instantiate since we don't know how to
    643     raise RuntimeError(msg) from None
--> 644 raise exception

TypeError: Caught TypeError in DataLoader worker process 0.
Original Traceback (most recent call last):
  File "/opt/conda/lib/python3.10/site-packages/torch/utils/data/_utils/worker.py", line 308, in _worker_loop
    data = fetcher.fetch(index)
  File "/opt/conda/lib/python3.10/site-packages/torch/utils/data/_utils/fetch.py", line 51, in fetch
    data = [self.dataset[idx] for idx in possibly_batched_index]
  File "/opt/conda/lib/python3.10/site-packages/torch/utils/data/_utils/fetch.py", line 51, in <listcomp>
    data = [self.dataset[idx] for idx in possibly_batched_index]
  File "/tmp/ipykernel_29/995576240.py", line 18, in __getitem__
    image = self.transforms(image=image)['image']
  File "/opt/conda/lib/python3.10/site-packages/albumentations/core/composition.py", line 210, in __call__
    data = t(**data)
  File "/opt/conda/lib/python3.10/site-packages/albumentations/core/transforms_interface.py", line 118, in __call__
    return self.apply_with_params(params, **kwargs)
  File "/opt/conda/lib/python3.10/site-packages/albumentations/core/transforms_interface.py", line 131, in apply_with_params
    res[key] = target_function(arg, **dict(params, **target_dependencies))
  File "/opt/conda/lib/python3.10/site-packages/albumentations/augmentations/transforms.py", line 213, in apply
    return F.normalize(image, self.mean, self.std, self.max_pixel_value)
  File "/opt/conda/lib/python3.10/site-packages/albumentations/augmentations/functional.py", line 89, in normalize
    mean = np.array(mean, dtype=np.float32)
  File "/opt/conda/lib/python3.10/site-packages/torch/_tensor.py", line 972, in __array__
    return self.numpy().astype(dtype, copy=False)
TypeError: can't convert cuda:0 device type tensor to numpy. Use Tensor.cpu() to copy the tensor to host memory first.
```

### Versions
```


### Steps to reproduce

### 🐛 Describe the bug
for code, please check hyperparameter section of https://www.kaggle.com/code/artemmalygin/pytorch-animals-10#Find-best-hyperparameters-with-optuna
```
def objective(trial):
    N_TRAIN_EXAMPLES = CONFIG['train_batch_size'] * 30
    N_VALID_EXAMPLES = CONFIG['valid_batch_size'] * 10
    device = CONFIG['device']

    # Generate the model.
    model = AnimalsModel(CONFIG['model_name'], labels)

    # Generate the optimizers.
    optimizer_number = trial.suggest_categorical('optimizer number (0: Adam; 1: SGD)', [0, 1])
    lr = trial.suggest_float('learning rate', 1e-2, 1e-1, log=True)
    
    optimizer_map = {
        0: 'Adam',
        1: 'SGD',
    }
    
    optimizer = getattr(optim, optimizer_map[optimizer_number])(model.parameters(), lr=lr)
    
    config = {
        'optimizer (0: Adam; 1: SGD)': optimizer_number,
        'learning rate': lr,
    }
    
    run = wandb.init(project='Animals10',
                     name=f'trial_{trial.number + 1}',
                     group='optuna research',
                     config=config,
                     anonymous='must')

    # Training of the model.
    for epoch in range(CONFIG['epochs']):
        model.train()
    
        dataset_size = 0
        running_loss = 0.0
        
        for batch_idx, data in enumerate(train_loader):
            # Limiting training data for faster epochs.
            if batch_idx * CONFIG['train_batch_size'] > N_TRAIN_EXAMPLES:
                break

            images = data['image'].to(device, dtype=torch.float)
            targets = data['target'].to(device, dtype=torch.long)
            targets = torch.reshape(targets, (-1,))

            batch_size = images.size(0)

            outputs = model(images)
            loss = criterion(outputs, targets)
            loss.backward()
            optimizer.step()
            
            # zero the parameter gradients
            optimizer.zero_grad()

        # Validation of the model.
        model.eval()
        correct = 0
        with torch.no_grad():
            for batch_idx, data in enumerate(valid_loader):
                # Limiting validation data.
                if batch_idx * CONFIG['valid_batch_size'] > N_VALID_EXAMPLES:
                    break
                
                images = data['image'].to(device, dtype=torch.float)
                targets = data['target'].to(device, dtype=torch.long)
                targets = torch.reshape(targets, (-1,))

                batch_size = images.size(0)

                outputs = model(images)
                loss = criterion(outputs, targets)

                running_loss += (loss.item() * batch_size)
                dataset_size += batch_size

        epoch_loss = running_loss / dataset_size
        run.log({'Cross Entropy Loss': epoch_loss})

        trial.report(epoch_loss, epoch)

        # handle pruning based on the intermediate value.
        if trial.should_prune():
            raise optuna.exceptions.TrialPruned()

    run.finish()
    
    return epoch_loss
```
```
study = optuna.create_study()
study.optimize(objective, n_trials=20, timeout=1800, show_progress_bar=True)

print('Number of finished trials: {}'.format(len(study.trials)))

print('Best trial:')
trial = study.best_trial

print('  Value: {}'.format(trial.value))

print('  Params: ')
for key, value in trial.params.items():
    print('    {}: {}'.format(key, value))
          
# Create the summary run.
summary = wandb.init(project='Animals10',
                     group='optuna summary',
                     name='summary')

# Getting the study trials.
trials = study.trials

# WandB summary.
for step, trial in enumerate(trials):
    # Logging the loss.
    summary.log({'Cross Entropy Loss': trial.value}, step=step, commit=False)

    # Logging the parameters.        
    summary.log(trial.params, commit=True)
    
summary.finish()
```
error:
```
View run at https://wandb.ai/andysingal/Animals10/runs/s26ixk76
[W 2023-09-09 14:57:30,926] Trial 0 failed with parameters: {'optimizer number (0: Adam; 1: SGD)': 0, 'learning rate': 0.0646134223274061} because of the following error: TypeError('Caught TypeError in DataLoader worker process 0.\nOriginal Traceback (most recent call last):\n  File "/opt/conda/lib/python3.10/site-packages/torch/utils/data/_utils/worker.py", line 308, in _worker_loop\n    data = fetcher.fetch(index)\n  File "/opt/conda/lib/python3.10/site-packages/torch/utils/data/_utils/fetch.py", line 51, in fetch\n    data = [self.dataset[idx] for idx in possibly_batched_index]\n  File "/opt/conda/lib/python3.10/site-packages/torch/utils/data/_utils/fetch.py", line 51, in <listcomp>\n    data = [self.dataset[idx] for idx in possibly_batched_index]\n  File "/tmp/ipykernel_29/995576240.py", line 18, in __getitem__\n    image = self.transforms(image=image)[\'image\']\n  File "/opt/conda/lib/python3.10/site-packages/albumentations/core/composition.py", line 210, in __call__\n    data = t(**data)\n  File "/opt/conda/lib/python3.10/site-packages/albumentations/core/transforms_interface.py", line 118, in __call__\n    return self.apply_with_params(params, **kwargs)\n  File "/opt/conda/lib/python3.10/site-packages/albumentations/core/transforms_interface.py", line 131, in apply_with_params\n    res[key] = target_function(arg, **dict(params, **target_dependencies))\n  File "/opt/conda/lib/python3.10/site-packages/albumentations/augmentations/transforms.py", line 213, in apply\n    return F.normalize(image, self.mean, self.std, self.max_pixel_value)\n  File "/opt/conda/lib/python3.10/site-packages/albumentations/augmentations/functional.py", line 89, in normalize\n    mean = np.array(mean, dtype=np.float32)\n  File "/opt/conda/lib/python3.10/site-packages/torch/_tensor.py", line 972, in __array__\n    return self.numpy().astype(dtype, copy=False)\nTypeError: can\'t convert cuda:0 device type tensor to numpy. Use Tensor.cpu() to copy the tensor to host memory first.\n').
Traceback (most recent call last):
  File "/opt/conda/lib/python3.10/site-packages/optuna/study/_optimize.py", line 200, in _run_trial
    value_or_values = func(trial)
  File "/tmp/ipykernel_29/2342615095.py", line 38, in objective
    for batch_idx, data in enumerate(train_loader):
  File "/opt/conda/lib/python3.10/site-packages/torch/utils/data/dataloader.py", line 634, in __next__
    data = self._next_data()
  File "/opt/conda/lib/python3.10/site-packages/torch/utils/data/dataloader.py", line 1346, in _next_data
    return self._process_data(data)
  File "/opt/conda/lib/python3.10/site-packages/torch/utils/data/dataloader.py", line 1372, in _process_data
    data.reraise()
  File "/opt/conda/lib/python3.10/site-packages/torch/_utils.py", line 644, in reraise
    raise exception
TypeError: Caught TypeError in DataLoader worker process 0.
Original Traceback (most recent call last):
  File "/opt/conda/lib/python3.10/site-packages/torch/utils/data/_utils/worker.py", line 308, in _worker_loop
    data = fetcher.fetch(index)
  File "/opt/conda/lib/python3.10/site-packages/torch/utils/data/_utils/fetch.py", line 51, in fetch
    data = [self.dataset[idx] for idx in possibly_batched_index]
  File "/opt/conda/lib/python3.10/site-packages/torch/utils/data/_utils/fetch.py", line 51, in <listcomp>
    data = [self.dataset[idx] for idx in possibly_batched_index]
  File "/tmp/ipykernel_29/995576240.py", line 18, in __getitem__
    image = self.transforms(image=image)['image']
  File "/opt/conda/lib/python3.10/site-packages/albumentations/core/composition.py", line 210, in __call__
    data = t(**data)
  File "/opt/conda/lib/python3.10/site-packages/albumentations/core/transforms_interface.py", line 118, in __call__
    return self.apply_with_params(params, **kwargs)
  File "/opt/conda/lib/python3.10/site-packages/albumentations/core/transforms_interface.py", line 131, in apply_with_params
    res[key] = target_function(arg, **dict(params, **target_dependencies))
  File "/opt/conda/lib/python3.10/site-packages/albumentations/augmentations/transforms.py", line 213, in apply
    return F.normalize(image, self.mean, self.std, self.max_pixel_value)
  File "/opt/conda/lib/python3.10/site-packages/albumentations/augmentations/functional.py", line 89, in normalize
    mean = np.array(mean, dtype=np.float32)
  File "/opt/conda/lib/python3.10/site-packages/torch/_tensor.py", line 972, in __array__
    return self.numpy().astype(dtype, copy=False)
TypeError: can't convert cuda:0 device type tensor to numpy. Use Tensor.cpu() to copy the tensor to host memory first.

[W 2023-09-09 14:57:30,942] Trial 0 failed with value None.
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
Cell In[49], line 2
      1 study = optuna.create_study()
----> 2 study.optimize(objective, n_trials=20, timeout=1800, show_progress_bar=True)
      4 print('Number of finished trials: {}'.format(len(study.trials)))
      6 print('Best trial:')

File /opt/conda/lib/python3.10/site-packages/optuna/study/study.py:442, in Study.optimize(self, func, n_trials, timeout, n_jobs, catch, callbacks, gc_after_trial, show_progress_bar)
    339 def optimize(
    340     self,
    341     func: ObjectiveFuncType,
   (...)
    348     show_progress_bar: bool = False,
    349 ) -> None:
    350     """Optimize an objective function.
    351 
    352     Optimization is done by choosing a suitable set of hyperparameter values from a given
   (...)
    440             If nested invocation of this method occurs.
    441     """
--> 442     _optimize(
    443         study=self,
    444         func=func,
    445         n_trials=n_trials,
    446         timeout=timeout,
    447         n_jobs=n_jobs,
    448         catch=tuple(catch) if isinstance(catch, Iterable) else (catch,),
    449         callbacks=callbacks,
    450         gc_after_trial=gc_after_trial,
    451         show_progress_bar=show_progress_bar,
    452     )

File /opt/conda/lib/python3.10/site-packages/optuna/study/_optimize.py:66, in _optimize(study, func, n_trials, timeout, n_jobs, catch, callbacks, gc_after_trial, show_progress_bar)
     64 try:
     65     if n_jobs == 1:
---> 66         _optimize_sequential(
     67             study,
     68             func,
     69             n_trials,
     70             timeout,
     71             catch,
     72             callbacks,
     73             gc_after_trial,
     74             reseed_sampler_rng=False,
     75             time_start=None,
     76             progress_bar=progress_bar,
     77         )
     78     else:
     79         if n_jobs == -1:

File /opt/conda/lib/python3.10/site-packages/optuna/study/_optimize.py:163, in _optimize_sequential(study, func, n_trials, timeout, catch, callbacks, gc_after_trial, reseed_sampler_rng, time_start, progress_bar)
    160         break
    162 try:
--> 163     frozen_trial = _run_trial(study, func, catch)
    164 finally:
    165     # The following line mitigates memory problems that can be occurred in some
    166     # environments (e.g., services that use computing containers such as GitHub Actions).
    167     # Please refer to the following PR for further details:
    168     # [https://github.com/optuna/optuna/pull/325.](https://github.com/optuna/optuna/pull/325.%3C/span%3E)
    169     if gc_after_trial:

File /opt/conda/lib/python3.10/site-packages/optuna/study/_optimize.py:251, in _run_trial(study, func, catch)
    244         assert False, "Should not reach."
    246 if (
    247     frozen_trial.state == TrialState.FAIL
    248     and func_err is not None
    249     and not isinstance(func_err, catch)
    250 ):
--> 251     raise func_err
    252 return frozen_trial

File /opt/conda/lib/python3.10/site-packages/optuna/study/_optimize.py:200, in _run_trial(study, func, catch)
    198 with get_heartbeat_thread(trial._trial_id, study._storage):
    199     try:
--> 200         value_or_values = func(trial)
    201     except exceptions.TrialPruned as e:
    202         # TODO(mamu): Handle multi-objective cases.
    203         state = TrialState.PRUNED

Cell In[48], line 38, in objective(trial)
     35 dataset_size = 0
     36 running_loss = 0.0
---> 38 for batch_idx, data in enumerate(train_loader):
     39     # Limiting training data for faster epochs.
     40     if batch_idx * CONFIG['train_batch_size'] > N_TRAIN_EXAMPLES:
     41         break

File /opt/conda/lib/python3.10/site-packages/torch/utils/data/dataloader.py:634, in _BaseDataLoaderIter.__next__(self)
    631 if self._sampler_iter is None:
    632     # TODO([https://github.com/pytorch/pytorch/issues/76750)](https://github.com/pytorch/pytorch/issues/76750)%3C/span%3E)
    633     self._reset()  # type: ignore[call-arg]
--> 634 data = self._next_data()
    635 self._num_yielded += 1
    636 if self._dataset_kind == _DatasetKind.Iterable and \
    637         self._IterableDataset_len_called is not None and \
    638         self._num_yielded > self._IterableDataset_len_called:

File /opt/conda/lib/python3.10/site-packages/torch/utils/data/dataloader.py:1346, in _MultiProcessingDataLoaderIter._next_data(self)
   1344 else:
   1345     del self._task_info[idx]
-> 1346     return self._process_data(data)

File /opt/conda/lib/python3.10/site-packages/torch/utils/data/dataloader.py:1372, in _MultiProcessingDataLoaderIter._process_data(self, data)
   1370 self._try_put_index()
   1371 if isinstance(data, ExceptionWrapper):
-> 1372     data.reraise()
   1373 return data

File /opt/conda/lib/python3.10/site-packages/torch/_utils.py:644, in ExceptionWrapper.reraise(self)
    640 except TypeError:
    641     # If the exception takes multiple arguments, don't try to
    642     # instantiate since we don't know how to
    643     raise RuntimeError(msg) from None
--> 644 raise exception

TypeError: Caught TypeError in DataLoader worker process 0.
Original Traceback (most recent call last):
  File "/opt/conda/lib/python3.10/site-packages/torch/utils/data/_utils/worker.py", line 308, in _worker_loop
    data = fetcher.fetch(index)
  File "/opt/conda/lib/python3.10/site-packages/torch/utils/data/_utils/fetch.py", line 51, in fetch
    data = [self.dataset[idx] for idx in possibly_batched_index]
  File "/opt/conda/lib/python3.10/site-packages/torch/utils/data/_utils/fetch.py", line 51, in <listcomp>
    data = [self.dataset[idx] for idx in possibly_batched_index]
  File "/tmp/ipykernel_29/995576240.py", line 18, in __getitem__
    image = self.transforms(image=image)['image']
  File "/opt/conda/lib/python3.10/site-packages/albumentations/core/composition.py", line 210, in __call__
    data = t(**data)
  File "/opt/conda/lib/python3.10/site-packages/albumentations/core/transforms_interface.py", line 118, in __call__
    return self.apply_with_params(params, **kwargs)
  File "/opt/conda/lib/python3.10/site-packages/albumentations/core/transforms_interface.py", line 131, in apply_with_params
    res[key] = target_function(arg, **dict(params, **target_dependencies))
  File "/opt/conda/lib/python3.10/site-packages/albumentations/augmentations/transforms.py", line 213, in apply
    return F.normalize(image, self.mean, self.std, self.max_pixel_value)
  File "/opt/conda/lib/python3.10/site-packages/albumentations/augmentations/functional.py", line 89, in normalize
    mean = np.array(mean, dtype=np.float32)
  File "/opt/conda/lib/python3.10/site-packages/torch/_tensor.py", line 972, in __array__
    return self.numpy().astype(dtype, copy=False)
TypeError: can't convert cuda:0 device type tensor to numpy. Use Tensor.cpu() to copy the tensor to host memory first.
```

### Versions


### Additional context (optional)

_No response_

**Comment by nzw0301** [comment]
The error message says 
>  TypeError: can't convert cuda:0 device type tensor to numpy. Use Tensor.cpu() to copy the tensor to host memory first.
